tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -812,34 +812,31 @@ ts_decompress_chunk_generate_paths(PlannerInfo *root, RelOptInfo *chunk_rel, Hyp
 			 */
 			if (batch_merge_path != NULL)
 			{
-				Path *merge_append_path =
-					(Path *) create_merge_append_path_compat(root,
-															 chunk_rel,
-															 list_make2(batch_merge_path,
-																		uncompressed_path),
-															 root->query_pathkeys,
-															 req_outer,
-															 NIL);
-
-				add_path(chunk_rel, merge_append_path);
+				path = (Path *) create_merge_append_path_compat(root,
+																chunk_rel,
+																list_make2(batch_merge_path,
+																		   uncompressed_path),
+																root->query_pathkeys,
+																req_outer,
+																NIL);
 			}
-
-			/*
-			 * Ideally, we would like for this to be a MergeAppend path.
-			 * However, accumulate_append_subpath will cut out MergeAppend
-			 * and directly add its children, so we have to combine the children
-			 * into a MergeAppend node later, at the chunk append level.
-			 */
-			path = (Path *) create_append_path_compat(root,
-													  chunk_rel,
-													  list_make2(path, uncompressed_path),
-													  NIL /* partial paths */,
-													  NIL /* pathkeys */,
-													  req_outer,
-													  0,
-													  false,
-													  false,
-													  path->rows + uncompressed_path->rows);
+			else
```

Comment:
Overall, the PR looks good. 

However, I have a question about this change. Why don't we add both paths (merge append of `batch_merge_path` and `uncompressed_path` and merge append of `path` and `uncompressed_path` ) and let PostgreSQL choose the cheapest one?

Reply:
thank you for the quick review! I did this because a tsbench run showed that without the else, both paths were added but the merged batch optimization was not picked, when in fact applying it would have been a better choice (53 ms vs 1415 ms for the query `SELECT * FROM ht_metrics_compressed ORDER BY time DESC LIMIT 1;` when all the chunks are marked as partially compressed). The big difference was due to sort pushdown below decompressChunk.

Reply:
But maybe what I should do here is examine the costs and adjust the costs instead.

Reply:
Thanks for the explanation. Do you remember the planned costs with and without the optimization (or the full query plans)? It might be interesting to see why this optimization is not picked up properly.

Reply:
query plan without the optimization: 
```
 Limit  (cost=115.22..115.23 rows=1 width=20) (actual time=1386.466..1386.472 rows=1 loops=1)
   Output: ht_metrics_compressed."time", ht_metrics_compressed.device, ht_metrics_compressed.value
   ->  Custom Scan (ChunkAppend) on public.ht_metrics_compressed  (cost=115.22..51305.24 rows=5119001 width=20) (actual time=1386.463..1386.468 rows=1 loops=1)
         Output: ht_metrics_compressed."time", ht_metrics_compressed.device, ht_metrics_compressed.value
         Order: ht_metrics_compressed."time" DESC
         Startup Exclusion: false
         Runtime Exclusion: false
         ->  Merge Append  (cost=115.22..51305.24 rows=5119001 width=20) (actual time=1386.461..1386.461 rows=1 loops=1)
               Sort Key: _hyper_4_15_chunk."time" DESC
               ->  Sort  (cost=25710.19..38507.69 rows=5119000 width=20) (actual time=1386.448..1386.448 rows=1 loops=1)
                     Output: _hyper_4_15_chunk."time", _hyper_4_15_chunk.device, _hyper_4_15_chunk.value
                     Sort Key: _hyper_4_15_chunk."time" DESC
                     Sort Method: top-N heapsort  Memory: 25kB
                     ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_15_chunk  (cost=0.02..115.19 rows=5119000 width=20) (actual time=0.353..593.148 rows=5112900 loops=1)
                           Output: _hyper_4_15_chunk."time", _hyper_4_15_chunk.device, _hyper_4_15_chunk.value
                           Bulk Decompression: true
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_20_chunk  (cost=0.00..115.19 rows=5119 width=88) (actual time=0.013..0.509 rows=5119 loops=1)
                                 Output: compress_hyper_5_20_chunk."time", compress_hyper_5_20_chunk.device, compress_hyper_5_20_chunk.value, compress_hyper_5_20_chunk._ts_meta_count, compress_hyper_5_20_chunk._ts_meta_sequence_num, compress_hyper_5_20_chunk._ts_meta_min_1, compress_hyper_5_20_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (actual time=0.010..0.010 rows=0 loops=1)
                     Output: _hyper_4_15_chunk."time", _hyper_4_15_chunk.device, _hyper_4_15_chunk.value
                     Sort Key: _hyper_4_15_chunk."time" DESC
                     Sort Method: quicksort  Memory: 25kB
                     ->  Seq Scan on _timescaledb_internal._hyper_4_15_chunk  (cost=0.00..0.00 rows=1 width=20) (actual time=0.006..0.006 rows=0 loops=1)
                           Output: _hyper_4_15_chunk."time", _hyper_4_15_chunk.device, _hyper_4_15_chunk.value
         ->  Merge Append  (cost=135.54..60645.56 rows=6051001 width=20) (never executed)
               Sort Key: _hyper_4_14_chunk."time" DESC
               ->  Sort  (cost=30390.51..45518.01 rows=6051000 width=20) (never executed)
                     Output: _hyper_4_14_chunk."time", _hyper_4_14_chunk.device, _hyper_4_14_chunk.value
                     Sort Key: _hyper_4_14_chunk."time" DESC
                     ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_14_chunk  (cost=0.02..135.51 rows=6051000 width=20) (never executed)
                           Output: _hyper_4_14_chunk."time", _hyper_4_14_chunk.device, _hyper_4_14_chunk.value
                           Bulk Decompression: false
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_19_chunk  (cost=0.00..135.51 rows=6051 width=88) (never executed)
                                 Output: compress_hyper_5_19_chunk."time", compress_hyper_5_19_chunk.device, compress_hyper_5_19_chunk.value, compress_hyper_5_19_chunk._ts_meta_count, compress_hyper_5_19_chunk._ts_meta_sequence_num, compress_hyper_5_19_chunk._ts_meta_min_1, compress_hyper_5_19_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_14_chunk."time", _hyper_4_14_chunk.device, _hyper_4_14_chunk.value
                     Sort Key: _hyper_4_14_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_14_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_14_chunk."time", _hyper_4_14_chunk.device, _hyper_4_14_chunk.value
         ->  Merge Append  (cost=135.53..60635.55 rows=6050001 width=20) (never executed)
               Sort Key: _hyper_4_13_chunk."time" DESC
               ->  Sort  (cost=30385.50..45510.50 rows=6050000 width=20) (never executed)
                     Output: _hyper_4_13_chunk."time", _hyper_4_13_chunk.device, _hyper_4_13_chunk.value
                     Sort Key: _hyper_4_13_chunk."time" DESC
                     ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_13_chunk  (cost=0.02..135.50 rows=6050000 width=20) (never executed)
                           Output: _hyper_4_13_chunk."time", _hyper_4_13_chunk.device, _hyper_4_13_chunk.value
                           Bulk Decompression: false
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_18_chunk  (cost=0.00..135.50 rows=6050 width=88) (never executed)
                                 Output: compress_hyper_5_18_chunk."time", compress_hyper_5_18_chunk.device, compress_hyper_5_18_chunk.value, compress_hyper_5_18_chunk._ts_meta_count, compress_hyper_5_18_chunk._ts_meta_sequence_num, compress_hyper_5_18_chunk._ts_meta_min_1, compress_hyper_5_18_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_13_chunk."time", _hyper_4_13_chunk.device, _hyper_4_13_chunk.value
                     Sort Key: _hyper_4_13_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_13_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_13_chunk."time", _hyper_4_13_chunk.device, _hyper_4_13_chunk.value
         ->  Merge Append  (cost=261.53..60761.55 rows=6050001 width=20) (never executed)
               Sort Key: _hyper_4_12_chunk."time" DESC
               ->  Sort  (cost=30511.50..45636.50 rows=6050000 width=20) (never executed)
                     Output: _hyper_4_12_chunk."time", _hyper_4_12_chunk.device, _hyper_4_12_chunk.value
                     Sort Key: _hyper_4_12_chunk."time" DESC
                     ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_12_chunk  (cost=0.04..261.50 rows=6050000 width=20) (never executed)
                           Output: _hyper_4_12_chunk."time", _hyper_4_12_chunk.device, _hyper_4_12_chunk.value
                           Bulk Decompression: false
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_17_chunk  (cost=0.00..261.50 rows=6050 width=88) (never executed)
                                 Output: compress_hyper_5_17_chunk."time", compress_hyper_5_17_chunk.device, compress_hyper_5_17_chunk.value, compress_hyper_5_17_chunk._ts_meta_count, compress_hyper_5_17_chunk._ts_meta_sequence_num, compress_hyper_5_17_chunk._ts_meta_min_1, compress_hyper_5_17_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_12_chunk."time", _hyper_4_12_chunk.device, _hyper_4_12_chunk.value
                     Sort Key: _hyper_4_12_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_12_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_12_chunk."time", _hyper_4_12_chunk.device, _hyper_4_12_chunk.value
         ->  Merge Append  (cost=3.83..803.85 rows=80001 width=20) (never executed)
               Sort Key: _hyper_4_11_chunk."time" DESC
               ->  Sort  (cost=403.80..603.80 rows=80000 width=20) (never executed)
                     Output: _hyper_4_11_chunk."time", _hyper_4_11_chunk.device, _hyper_4_11_chunk.value
                     Sort Key: _hyper_4_11_chunk."time" DESC
                     ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_11_chunk  (cost=0.05..3.80 rows=80000 width=20) (never executed)
                           Output: _hyper_4_11_chunk."time", _hyper_4_11_chunk.device, _hyper_4_11_chunk.value
                           Bulk Decompression: false
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_16_chunk  (cost=0.00..3.80 rows=80 width=88) (never executed)
                                 Output: compress_hyper_5_16_chunk."time", compress_hyper_5_16_chunk.device, compress_hyper_5_16_chunk.value, compress_hyper_5_16_chunk._ts_meta_count, compress_hyper_5_16_chunk._ts_meta_sequence_num, compress_hyper_5_16_chunk._ts_meta_min_1, compress_hyper_5_16_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_11_chunk."time", _hyper_4_11_chunk.device, _hyper_4_11_chunk.value
                     Sort Key: _hyper_4_11_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_11_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_11_chunk."time", _hyper_4_11_chunk.device, _hyper_4_11_chunk.value
 Query Identifier: -2027753307672493323
 Planning Time: 3.100 ms
 Execution Time: 1386.662 ms
 ```
 and query plan with the `decompression_sorted_merge` on:
 ```
  Limit  (cost=443.38..447.49 rows=1 width=20) (actual time=7.994..8.010 rows=1 loops=1)
   Output: ht_metrics_compressed."time", ht_metrics_compressed.device, ht_metrics_compressed.value
   ->  Custom Scan (ChunkAppend) on public.ht_metrics_compressed  (cost=443.38..21014962.20 rows=5119001 width=20) (actual time=7.990..8.003 rows=1 loops=1)
         Output: ht_metrics_compressed."time", ht_metrics_compressed.device, ht_metrics_compressed.value
         Order: ht_metrics_compressed."time" DESC
         Startup Exclusion: false
         Runtime Exclusion: false
         ->  Merge Append  (cost=443.38..21014962.20 rows=5119001 width=20) (actual time=7.988..7.989 rows=1 loops=1)
               Sort Key: _hyper_4_15_chunk."time" DESC
               ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_15_chunk  (cost=443.36..20963772.16 rows=5119000 width=20) (actual time=7.947..7.948 rows=1 loops=1)
                     Output: _hyper_4_15_chunk."time", _hyper_4_15_chunk.device, _hyper_4_15_chunk.value
                     Sorted merge append: true
                     Bulk Decompression: false
                     ->  Sort  (cost=0.00..0.00 rows=0 width=0) (actual time=7.553..7.554 rows=2 loops=1)
                           Output: compress_hyper_5_20_chunk."time", compress_hyper_5_20_chunk.device, compress_hyper_5_20_chunk.value, compress_hyper_5_20_chunk._ts_meta_count, compress_hyper_5_20_chunk._ts_meta_sequence_num, compress_hyper_5_20_chunk._ts_meta_min_1, compress_hyper_5_20_chunk._ts_meta_max_1
                           Sort Key: compress_hyper_5_20_chunk._ts_meta_max_1 DESC
                           Sort Method: quicksort  Memory: 797kB
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_20_chunk  (cost=0.00..115.19 rows=5119 width=88) (actual time=0.018..1.540 rows=5119 loops=1)
                                 Output: compress_hyper_5_20_chunk."time", compress_hyper_5_20_chunk.device, compress_hyper_5_20_chunk.value, compress_hyper_5_20_chunk._ts_meta_count, compress_hyper_5_20_chunk._ts_meta_sequence_num, compress_hyper_5_20_chunk._ts_meta_min_1, compress_hyper_5_20_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (actual time=0.038..0.038 rows=0 loops=1)
                     Output: _hyper_4_15_chunk."time", _hyper_4_15_chunk.device, _hyper_4_15_chunk.value
                     Sort Key: _hyper_4_15_chunk."time" DESC
                     Sort Method: quicksort  Memory: 25kB
                     ->  Seq Scan on _timescaledb_internal._hyper_4_15_chunk  (cost=0.00..0.00 rows=1 width=20) (actual time=0.021..0.021 rows=0 loops=1)
                           Output: _hyper_4_15_chunk."time", _hyper_4_15_chunk.device, _hyper_4_15_chunk.value
         ->  Merge Append  (cost=530.75..29352721.56 rows=6051001 width=20) (never executed)
               Sort Key: _hyper_4_14_chunk."time" DESC
               ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_14_chunk  (cost=530.73..29292211.53 rows=6051000 width=20) (never executed)
                     Output: _hyper_4_14_chunk."time", _hyper_4_14_chunk.device, _hyper_4_14_chunk.value
                     Sorted merge append: true
                     Bulk Decompression: false
                     ->  Sort  (cost=0.00..0.00 rows=0 width=0) (never executed)
                           Output: compress_hyper_5_19_chunk."time", compress_hyper_5_19_chunk.device, compress_hyper_5_19_chunk.value, compress_hyper_5_19_chunk._ts_meta_count, compress_hyper_5_19_chunk._ts_meta_sequence_num, compress_hyper_5_19_chunk._ts_meta_min_1, compress_hyper_5_19_chunk._ts_meta_max_1
                           Sort Key: compress_hyper_5_19_chunk._ts_meta_max_1 DESC
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_19_chunk  (cost=0.00..135.51 rows=6051 width=88) (never executed)
                                 Output: compress_hyper_5_19_chunk."time", compress_hyper_5_19_chunk.device, compress_hyper_5_19_chunk.value, compress_hyper_5_19_chunk._ts_meta_count, compress_hyper_5_19_chunk._ts_meta_sequence_num, compress_hyper_5_19_chunk._ts_meta_min_1, compress_hyper_5_19_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_14_chunk."time", _hyper_4_14_chunk.device, _hyper_4_14_chunk.value
                     Sort Key: _hyper_4_14_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_14_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_14_chunk."time", _hyper_4_14_chunk.device, _hyper_4_14_chunk.value
         ->  Merge Append  (cost=530.67..29343030.68 rows=6050001 width=20) (never executed)
               Sort Key: _hyper_4_13_chunk."time" DESC
               ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_13_chunk  (cost=530.65..29282530.65 rows=6050000 width=20) (never executed)
                     Output: _hyper_4_13_chunk."time", _hyper_4_13_chunk.device, _hyper_4_13_chunk.value
                     Sorted merge append: true
                     Bulk Decompression: false
                     ->  Sort  (cost=0.00..0.00 rows=0 width=0) (never executed)
                           Output: compress_hyper_5_18_chunk."time", compress_hyper_5_18_chunk.device, compress_hyper_5_18_chunk.value, compress_hyper_5_18_chunk._ts_meta_count, compress_hyper_5_18_chunk._ts_meta_sequence_num, compress_hyper_5_18_chunk._ts_meta_min_1, compress_hyper_5_18_chunk._ts_meta_max_1
                           Sort Key: compress_hyper_5_18_chunk._ts_meta_max_1 DESC
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_18_chunk  (cost=0.00..135.50 rows=6050 width=88) (never executed)
                                 Output: compress_hyper_5_18_chunk."time", compress_hyper_5_18_chunk.device, compress_hyper_5_18_chunk.value, compress_hyper_5_18_chunk._ts_meta_count, compress_hyper_5_18_chunk._ts_meta_sequence_num, compress_hyper_5_18_chunk._ts_meta_min_1, compress_hyper_5_18_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_13_chunk."time", _hyper_4_13_chunk.device, _hyper_4_13_chunk.value
                     Sort Key: _hyper_4_13_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_13_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_13_chunk."time", _hyper_4_13_chunk.device, _hyper_4_13_chunk.value
         ->  Merge Append  (cost=656.67..29343156.68 rows=6050001 width=20) (never executed)
               Sort Key: _hyper_4_12_chunk."time" DESC
               ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_12_chunk  (cost=656.65..29282656.65 rows=6050000 width=20) (never executed)
                     Output: _hyper_4_12_chunk."time", _hyper_4_12_chunk.device, _hyper_4_12_chunk.value
                     Sorted merge append: true
                     Bulk Decompression: false
                     ->  Sort  (cost=0.00..0.00 rows=0 width=0) (never executed)
                           Output: compress_hyper_5_17_chunk."time", compress_hyper_5_17_chunk.device, compress_hyper_5_17_chunk.value, compress_hyper_5_17_chunk._ts_meta_count, compress_hyper_5_17_chunk._ts_meta_sequence_num, compress_hyper_5_17_chunk._ts_meta_min_1, compress_hyper_5_17_chunk._ts_meta_max_1
                           Sort Key: compress_hyper_5_17_chunk._ts_meta_max_1 DESC
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_17_chunk  (cost=0.00..261.50 rows=6050 width=88) (never executed)
                                 Output: compress_hyper_5_17_chunk."time", compress_hyper_5_17_chunk.device, compress_hyper_5_17_chunk.value, compress_hyper_5_17_chunk._ts_meta_count, compress_hyper_5_17_chunk._ts_meta_sequence_num, compress_hyper_5_17_chunk._ts_meta_min_1, compress_hyper_5_17_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_12_chunk."time", _hyper_4_12_chunk.device, _hyper_4_12_chunk.value
                     Sort Key: _hyper_4_12_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_12_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_12_chunk."time", _hyper_4_12_chunk.device, _hyper_4_12_chunk.value
         ->  Merge Append  (cost=6.55..5926.56 rows=80001 width=20) (never executed)
               Sort Key: _hyper_4_11_chunk."time" DESC
               ->  Custom Scan (DecompressChunk) on _timescaledb_internal._hyper_4_11_chunk  (cost=6.53..5126.53 rows=80000 width=20) (never executed)
                     Output: _hyper_4_11_chunk."time", _hyper_4_11_chunk.device, _hyper_4_11_chunk.value
                     Sorted merge append: true
                     Bulk Decompression: false
                     ->  Sort  (cost=0.00..0.00 rows=0 width=0) (never executed)
                           Output: compress_hyper_5_16_chunk."time", compress_hyper_5_16_chunk.device, compress_hyper_5_16_chunk.value, compress_hyper_5_16_chunk._ts_meta_count, compress_hyper_5_16_chunk._ts_meta_sequence_num, compress_hyper_5_16_chunk._ts_meta_min_1, compress_hyper_5_16_chunk._ts_meta_max_1
                           Sort Key: compress_hyper_5_16_chunk._ts_meta_max_1 DESC
                           ->  Seq Scan on _timescaledb_internal.compress_hyper_5_16_chunk  (cost=0.00..3.80 rows=80 width=88) (never executed)
                                 Output: compress_hyper_5_16_chunk."time", compress_hyper_5_16_chunk.device, compress_hyper_5_16_chunk.value, compress_hyper_5_16_chunk._ts_meta_count, compress_hyper_5_16_chunk._ts_meta_sequence_num, compress_hyper_5_16_chunk._ts_meta_min_1, compress_hyper_5_16_chunk._ts_meta_max_1
               ->  Sort  (cost=0.01..0.02 rows=1 width=20) (never executed)
                     Output: _hyper_4_11_chunk."time", _hyper_4_11_chunk.device, _hyper_4_11_chunk.value
                     Sort Key: _hyper_4_11_chunk."time" DESC
                     ->  Seq Scan on _timescaledb_internal._hyper_4_11_chunk  (cost=0.00..0.00 rows=1 width=20) (never executed)
                           Output: _hyper_4_11_chunk."time", _hyper_4_11_chunk.device, _hyper_4_11_chunk.value
 Query Identifier: -2027753307672493323
 Planning Time: 4.289 ms
 Execution Time: 8.727 ms
(92 rows)
```